lib/Dialect/Moore/MooreOps.cpp

```diff
@@ -507,121 +516,96 @@ LogicalResult ConcatRefOp::inferReturnTypes(
 // StructCreateOp
 //===----------------------------------------------------------------------===//
 
-LogicalResult StructCreateOp::verify() {
-  /// checks if the types of the inputs are exactly equal to the types of the
-  /// result struct fields
-  return TypeSwitch<Type, LogicalResult>(getType().getNestedType())
-      .Case<StructType, UnpackedStructType>([this](auto &type) {
-        auto members = type.getMembers();
-        auto inputs = getInput();
-        if (inputs.size() != members.size())
-          return failure();
-        for (size_t i = 0; i < members.size(); i++) {
-          auto memberType = cast<UnpackedType>(members[i].type);
-          auto inputType = inputs[i].getType();
-          if (inputType != memberType) {
-            emitOpError("input types must match struct field types and orders");
-            return failure();
-          }
-        }
-        return success();
-      })
-      .Default([this](auto &) {
-        emitOpError("Result type must be StructType or UnpackedStructType");
-        return failure();
-      });
+static std::optional<uint32_t> getStructFieldIndex(Type type, StringAttr name) {
+  if (auto structType = dyn_cast<StructType>(type))
+    return structType.getFieldIndex(name);
+  if (auto structType = dyn_cast<UnpackedStructType>(type))
+    return structType.getFieldIndex(name);
+  assert(0 && "expected StructType or UnpackedStructType");
+  return {};
 }
 
-OpFoldResult StructCreateOp::fold(FoldAdaptor adaptor) {
-  auto inputs = adaptor.getInput();
+static ArrayRef<StructLikeMember> getStructMembers(Type type) {
+  if (auto structType = dyn_cast<StructType>(type))
+    return structType.getMembers();
+  if (auto structType = dyn_cast<UnpackedStructType>(type))
+    return structType.getMembers();
+  assert(0 && "expected StructType or UnpackedStructType");
+  return {};
+}
 
-  if (llvm::any_of(inputs, [](Attribute attr) { return !attr; }))
-    return {};
+static UnpackedType getStructFieldType(Type type, StringAttr name) {
+  if (auto index = getStructFieldIndex(type, name))
+    return getStructMembers(type)[*index].type;
+  return {};
+}
 
```

Comment:
Wonderful interface for `structLikeType`.  Will it be better if  let be Interfaces for `StructLikeType` which means moving these to `mooreTypes.cpp`?

Reply:
I'm currently experimenting with creating a common base class for `StructType` and `UnpackedStructType`, such that you can do a `dyn_cast<AnyStructType>(type)` to go from `StructType` and `UnpackedStructType` to a common `AnyStructType` base class. Hopefully I can move this code into that base class in `MooreTypes.cpp` as you suggest :smiley: